### PR TITLE
GO-2973 Integrate tantivy-go

### PR DESCRIPTION
**Pull request by fat-fellow**
---

- [ ] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: pkg/lib/localstore/ftsearch/autobatchertantivy.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	}
+}
+
+func (f *ftSearch2) Iterate(objectId string, fields []string, shouldContinue func(doc *SearchDoc) bool) (err error) {
+	result, err := f.index.Search(fmt.Sprintf("%s:%s", fieldId, objectId), docLimit, false, fieldId)
+	if err != nil {
+		return err
+	}
+
+	var p fastjson.Parser
```
Comment (deff7) [inline pkg/lib/localstore/ftsearch/autobatchertantivy.go:25]: you also could use https://pkg.go.dev/github.com/valyala/fastjson#ParserPool

Comment (fat-fellow) [reply pkg/lib/localstore/ftsearch/autobatchertantivy.go:25]: done

File: pkg/lib/localstore/ftsearch/ftsearchtantivy.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+	f.index.Free()
+	return nil
+}
+
+func (f *ftSearch2) cleanupBleve() {
+	_ = os.RemoveAll(filepath.Join(f.rootPath, ftsDir))
+}
+
+func escapeQuery(query string) string {
+	specialChars := []rune{'+', '-', '&', '|', '!', '(', ')', '{', '}', '[', ']', '^', '"', '~', '*', '?', ':'}
```
Comment (deff7) [inline pkg/lib/localstore/ftsearch/ftsearchtantivy.go:356]: use map

File: pkg/lib/localstore/objectstore/queries.go [hunk 3]
```diff
@@ -423,34 +436,30 @@ func (s *dsObjectStore) performFulltextSearch(text string, highlightFormatter ft
...
+	return nil
 }
 
-func iterateOverAndFilters(fs []database.Filter) (spaceID string) {
+func iterateOverAndFilters(fs []database.Filter) []string {
+	var spaceIds []string
 	for _, f := range fs {
-		if spaceID = getSpaceIDFromFilter(f); spaceID != "" {
-			return spaceID
+		if newSpaceIds := getSpaceIdsFromFilter(f); len(newSpaceIds) != 0 {
```
Comment (deff7) [inline pkg/lib/localstore/objectstore/queries.go:458]: just 

```
spaceIds = append(spaceIds, getSpaceIdsFromFilter(f)...)
```

File: pkg/lib/localstore/ftsearch/ftsearchbleve.go [hunk 4]
```diff
@@ -96,6 +98,7 @@ func (f *ftSearch) Run(context.Context) (err error) {
 		return
 	}
 	f.index = index
+	f.cleanTantivy()
```
[the remark this answers is not in the document]
Comment (fat-fellow) [reply pkg/lib/localstore/ftsearch/ftsearchbleve.go:101]: Nope, it is for switching between the engines

**Review by deff7 — APPROVED** [review]

**Pull request merged**